src/bio/bio_device.c

```diff
@@ -456,3 +456,110 @@ bio_replace_dev(struct bio_xs_context *xs_ctxt, uuid_t old_dev_id,
 		smd_free_dev_info(new_info);
 	return rc;
 }
+
+static struct bio_dev_info *
+alloc_dev_info(uuid_t dev_id, struct smd_dev_info *s_info)
+{
+	struct bio_dev_info	*info;
+	int			 tgt_cnt = 0, i;
+
+	D_ALLOC_PTR(info);
+	if (info == NULL)
+		return NULL;
+
+	if (s_info != NULL) {
+		tgt_cnt = s_info->sdi_tgt_cnt;
+		info->bdi_flags |= NVME_DEV_FL_INUSE;
+		if (s_info->sdi_state == SMD_DEV_FAULTY)
+			info->bdi_flags |= NVME_DEV_FL_FAULTY;
+	}
+
+	if (tgt_cnt != 0) {
+		D_ALLOC_ARRAY(info->bdi_tgts, tgt_cnt);
+		if (info->bdi_tgts == NULL) {
+			D_FREE(info);
+			return NULL;
+		}
+	}
+
+	D_INIT_LIST_HEAD(&info->bdi_link);
+	uuid_copy(info->bdi_dev_id, dev_id);
+	info->bdi_tgt_cnt = tgt_cnt;
+	for (i = 0; i < info->bdi_tgt_cnt; i++)
+		info->bdi_tgts[i] = s_info->sdi_tgts[i];
+
+	return info;
+}
+
+static struct smd_dev_info *
+find_smd_dev(uuid_t dev_id, d_list_t *s_dev_list)
+{
+	struct smd_dev_info	*s_info;
+
+	d_list_for_each_entry(s_info, s_dev_list, sdi_link) {
+		if (uuid_compare(s_info->sdi_id, dev_id) == 0)
+			return s_info;
+	}
+
+	return NULL;
+}
+
+int
+bio_dev_list(struct bio_xs_context *xs_ctxt, d_list_t *dev_list)
+{
+	d_list_t		 s_dev_list;
+	struct bio_dev_info	*b_info;
+	struct smd_dev_info	*s_info, *s_tmp;
+	struct bio_bdev		*d_bdev;
+	int			 rc, dev_cnt = 0;
+
+	/* Caller ensures the request handling ULT created on init xstream */
+	D_ASSERT(is_init_xstream(xs_ctxt));
+
+	D_ASSERT(dev_list != NULL && d_list_empty(dev_list));
+	D_INIT_LIST_HEAD(&s_dev_list);
+
+	rc = smd_dev_list(&s_dev_list, &dev_cnt);
+	if (rc) {
+		D_ERROR("Failed to get SMD dev list "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	/* Scan all plugged devices */
```

Comment:
so this will scan all of the bio bdevs to find matching SMD entries, and then once found delete the SMD entry? this is not clear to me, seems we have 3 cases in which we need to return all devices from the bio_dev_list call:
1. New device - bio_bdev_list() will contain this device, but device will not be present in SMD dev list. In this case we need to call alloc_dev_info and add the device to the returned dev_list
2. In Use Device - bio_bdev_list() will contain this device and so will smd dev list. In this case we would call alloc_dev_info and again add the device to the returned dev_list.
3. Removed device - bio_bdev_list() will not contain the device but smd dev list will. I see this functionality represented on line 548 were all remaining SMD devices are scanned. This is the only case were we should remove from the smd dev list and not add anything to the returned dev_list.

I am working on the patch now to call bio_device_list() from dmg storage query list-devices command. This command should return only all plugged in devices right (ie NEW, NORMAL, or FAULTY). 

maybe this logic is present here, but it is just not currently clear to me.

Reply:
bio_dev_list() will return all devices, including the unplugged devices (only presented in SMD), each device is represented by a bio_dev_info, and caller can tell if the device is plugged, inuse or faulty by the 'bdi_flags'.